### ISSUE TITLE
False Positive: `no-member` for `ParamSpec` under Python 3.12

### ISSUE THREAD

**Issue by lord-haffi** [pylint-dev/pylint#9401]
### Bug description

A piece of code which worked fine for Python 3.11 (pylint v3.0.3 doesn't report anything here). But with the upgrade to 3.12 `pylint` suddenly doesn't know the members of `ParamSpec` anymore. I know that the generic typing got reworked in 3.12, maybe I'm doing something wrong. I took a quick look at the `typing` package in 3.12 and couldn't find any declaration of `ParamSpec` so it might be more a problem of the Python 3.12 release.

```python
import inspect
from typing import Callable, TypeVar, ParamSpec

T = TypeVar("T")
P = ParamSpec("P")

def bind_arguments(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> inspect.BoundArguments:
    """
    Binds the given arguments to the given function. If the arguments don't match the functions signature, it will
    raise an error.
    """
    bound_arguments = inspect.signature(func).bind(*args, **kwargs)
    bound_arguments.apply_defaults()
    return bound_arguments
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a
a.py:5:48: E1101: Instance of 'ParamSpec' has no 'args' member (no-member)
a.py:5:66: E1101: Instance of 'ParamSpec' has no 'kwargs' member (no-member)
```

### Expected behavior

No false-positive here. `mypy` has no complains about this code and the tests work fine under Python 3.12 as well.

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.1
Python 3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]
```

### OS / Environment

Windows 11

**Comment by jacobtylerwalls** [comment]
Hi @lord-haffi, would you mind updating the snippet to include imports and assignments of P and T? Thanks!

**Comment by jacobtylerwalls** [comment]
Thanks for the report!

This diff to astroid should fix it:
```diff
diff --git a/astroid/brain/brain_typing.py b/astroid/brain/brain_typing.py
index ea496440..20276b6c 100644
--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -437,7 +437,13 @@ def _typing_transform():
     class Generic:
         @classmethod
         def __class_getitem__(cls, item):  return cls
-    class ParamSpec: ...
+    class ParamSpec:
+        @property
+        def args(self):
+            return ParamSpecArgs(self)
+        @property
+        def kwargs(self):
+            return ParamSpecKwargs(self)
     class ParamSpecArgs: ...
     class ParamSpecKwargs: ...
     class TypeAlias: ...
```

> I took a quick look at the typing package in 3.12 and couldn't find any declaration of ParamSpec

This is exactly pylint's issue. In 3.12 CPython decided to C-accelerate the typing module, and that hosed a lot of pylint's ability to inspect typing classes. I stubbed out the bare minimum in pylint-dev/astroid@f2120dbff587f9512f64b90a9d0f074542f3d317, but there's a lot missing.

**Comment by DanielNoord** [comment]
@jacobtylerwalls Is this the same issue as with `collections`? I thought we fixed that by making our import resolved accept the frozen modules?

**Comment by DanielNoord** [comment]
See https://github.com/pylint-dev/astroid/issues/1512 and related PRs. We might be able to come up with a similar fix here.

**Comment by lord-haffi** [comment]
> would you mind updating the snippet to include imports and assignments of P and T?

Sure! Well, you already figured it out but for the sake of completeness I edited my post.

Thanks!

**Comment by jacobtylerwalls** [comment]
@DanielNoord it's actually not as bad as the collections issue.

I looked into the reason that I [stubbed](https://github.com/pylint-dev/astroid/commit/f2120dbff587f9512f64b90a9d0f074542f3d317) out the typing module. It turns out it was really just to pass about a dozen or so cases in astroid that depend on being able to infer the bases of a class using subscript syntax. It doesn't yet work with the new PEP 695 syntax. We should fix it in a minor release when someone volunteers. In the meantime, I think my stopgap is good for a patch release.

I'll open an astroid issue with my finding.

**Comment by jacobtylerwalls** [comment]
Opened https://github.com/pylint-dev/astroid/issues/2370 for the real fix. Will prepare a patch with the band-aid.
